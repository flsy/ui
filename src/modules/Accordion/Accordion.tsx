import React, { ReactNode } from 'react';
import styled from 'styled-components';

interface IProps {
  className?: string;
  children: ReactNode;
}

const Accordion = ({ className, children }: IProps): JSX.Element => <div className={className}>{children}</div>;

export default styled(Accordion)`
  margin: 0 auto;
`;

Accordion.defaultProps = {
  className: undefined,
};
