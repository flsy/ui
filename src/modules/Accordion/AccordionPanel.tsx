import React, { ReactNode } from 'react';
import styled, { css } from 'styled-components';
import Button from '../Button/Button';
import DownIcon from '../Icon/DownIcon';
import UpIcon from '../Icon/UpIcon';

interface IProps {
  isOpen: boolean;
  title: string;
  onOpen: () => void;
  className?: string;
  onClose?: () => void;
  children?: ReactNode;
}

const Header = styled.div<{ isOpen: boolean }>`
  cursor: pointer;
  user-select: none;
  background-color: ${({ theme }) => theme.background.default};
  border: 1px solid ${({ theme }) => theme.colors.border};
  border-radius: ${({ theme }) => theme.borderRadius};
  padding: 8px;
  display: flex;
  justify-content: space-between;

  ${({ isOpen }) =>
    isOpen &&
    css`
      border-bottom: none;
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    `}
`;

const Content = styled.div`
  padding: 1em;
  margin: auto;
  border: 1px solid ${({ theme }) => theme.colors.border};
`;

const SmallButton = styled(Button)`
  font-size: 10px;
`;

const AccordionPanel = ({ className, title, isOpen, onClose, onOpen, children }: IProps) => (
  <div className={className}>
    <Header isOpen={isOpen} onClick={isOpen ? onClose : onOpen}>
      {title}
      {onClose && <SmallButton text="" link={true} onClick={isOpen ? onClose : onOpen} iconRight={isOpen ? <UpIcon /> : <DownIcon />} />}
    </Header>
    {isOpen && <Content>{children}</Content>}
  </div>
);

AccordionPanel.defaultProps = {
  className: undefined,
  onClose: undefined,
  children: undefined,
};

export default styled(AccordionPanel)`
  margin: 8px 4px;
`;
