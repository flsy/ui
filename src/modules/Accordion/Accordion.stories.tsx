import { storiesOf } from '@storybook/react';
import React from 'react';
import styled from 'styled-components';
import { Accordion, Container } from '../../index';
import AccordionPanel from './AccordionPanel';

const AccordionContent = ({ openId, setOpenId }) => (
  <>
    <AccordionPanel isOpen={openId === 0} title="Step 1" onOpen={() => setOpenId(0)} onClose={() => setOpenId()}>
      first content
    </AccordionPanel>
    <AccordionPanel isOpen={openId === 1} title="Step 2" onOpen={() => setOpenId(1)} onClose={() => setOpenId()}>
      second content
    </AccordionPanel>
    <AccordionPanel isOpen={openId === 2} title="Step 3" onOpen={() => setOpenId(2)} onClose={() => setOpenId()}>
      third content
    </AccordionPanel>
  </>
);

const AccordionPreview = () => {
  const [openId, setOpenId] = React.useState(0);
  return (
    <Container size="xl">
      <Accordion>
        <AccordionContent openId={openId} setOpenId={setOpenId} />
      </Accordion>
    </Container>
  );
};

const StyledAccordion = styled(Accordion)`
  color: red;
`;

const StyledAccordionPreview = () => {
  const [openId, setOpenId] = React.useState(0);
  return (
    <Container size="xl">
      <StyledAccordion>
        <AccordionContent openId={openId} setOpenId={setOpenId} />
      </StyledAccordion>
    </Container>
  );
};

storiesOf('Accordion', module)
  .add('basic usage', () => <AccordionPreview />)
  .add('styled', () => <StyledAccordionPreview />);
