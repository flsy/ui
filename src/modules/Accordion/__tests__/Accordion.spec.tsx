import { shallow } from 'enzyme';
import React from 'react';
import Accordion from '../Accordion';
import { AccordionPanel } from '../index';

describe('<Accordion />', () => {
  it('should render', () => {
    const accordion = shallow(
      <Accordion>
        <AccordionPanel isOpen={true} onOpen={() => {}} title="Panel title">
          Panel content
        </AccordionPanel>
      </Accordion>,
    );
    expect(accordion).toMatchSnapshot();
  });
  it('should render panel content', () => {
    const accordion = shallow(
      <Accordion>
        <AccordionPanel isOpen={true} onOpen={() => {}} title="Panel title">
          Panel content
        </AccordionPanel>
      </Accordion>,
    );
    expect(accordion.find(AccordionPanel).children().text()).toBe('Panel content');
  });
});
