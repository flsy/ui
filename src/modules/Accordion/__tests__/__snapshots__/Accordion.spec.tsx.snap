// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> should render 1`] = `
.c0 {
  margin: 0 auto;
}

<Accordion
  className="c0"
>
  <Styled(AccordionPanel)
    isOpen={true}
    onOpen={[Function]}
    title="Panel title"
  >
    Panel content
  </Styled(AccordionPanel)>
</Accordion>
`;
